src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -71,8 +71,13 @@ protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException
                 tmp.renameTo(target);
 
                 if (target.exists()) {
+                    Checksum expected = new Checksum(sum1, sum2);
+                    Checksum actual = Checksum.forFile(target);
+                    if (!expected.equals(actual)) {
+                        throw new IOException("Checksum of retrieved jar doesn't match: " + target);
+                    }
                     // even if we fail to rename, we are OK as long as the target actually exists at this point
-                    // this can happen if two FileSystejarCache instances share the same cache dir
+                    // this can happen if two FileSystemJarCache instances share the same cache dir
```

Comment:
It makes sense. Two instances can indeed share a cache directory. So a failure in `renameTo` is OK so long as the target actually exists _and_ the checksum is still correct.